### PR TITLE
Crowd-Strike FalconX sandbox

### PR DESCRIPTION
**Pull request by ohaim1008**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/19771

## Description
new integration for FalconX sandbox,
currently tested locally

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_description.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## Hello World
+- This text is markdown
+- Here you should explain how to configure the instance in Demisto
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_description.md:3]: update

File: Packs/CrowdStrikeFalconX/pack_metadata.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "## FILL OUT MANUALLY ##",
+    "description": "## FILL OUT MANUALLY ##",
+    "support": "demisto",
+    "serverMinVersion": "## FILL OUT MANUALLY #",
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/pack_metadata.json:5]: update. also the use cases, and it should be xsoar, not demisto :trynottocry:.

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 3]
```diff
@@ -0,0 +1,511 @@
+from typing import Dict, Callable, Tuple, Any
+
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+import json
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:7]: can rm, no use for this module

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+
+
+class Client(BaseClient):
+    """
+    Client to use in the CrowdStrikeFalconX integration. Uses BaseClient
+    """
+
+    def __init__(self, server_url: str, username: str, password: str, use_ssl: bool):
+        super().__init__(base_url=server_url, verify=use_ssl)
+        self._base_url = server_url
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:39]: no need, you create it in the super()

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+    return group_outputs
+
+
+def parse_outputs(
+        api_res: Dict,
+        meta_fields: list = [],
+        quota_fields: list = [],
+        resources_fields: list = [],
+        sandbox_filds: list = []
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:98]:
```suggestion
        sandbox_fileds: list = []
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+    if api_res.get("resources"):
+        # depended on the command, the resources section can be a str list or a list that contains
+        # only one argument which is a dict
+        if type(api_res.get("resources")[0]) == dict:
+            api_res_resources = api_res.get("resources")[0]
+            resources_group_outputs = add_outputs_from_dict(api_res_resources, resources_fields)
+
+            if api_res_resources and api_res_resources.get("sandbox"):
+                api_res_sandbox = api_res_resources.get("sandbox")[0]
+                sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, sandbox_filds)
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:135]:
```suggestion
                sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, sandbox_fields)
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+        else:
+            # the resources section is a list of strings
+            resources_group_outputs = {"resources": api_res.get("resources")}
+
+    merged_dicts = {**meta_group_outputs, **quota_group_outputs, **resources_group_outputs, **sandbox_group_outputs}
+
+    return {f'csfalconx.resource(val.resource === obj.resource)': merged_dicts}
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:142]: I'm not sure that this is how the merging DT works. please try it. IMO it's something like:
```suggestion
    return {f'csfalconx.resource(val.id === obj.id)': merged_dicts}
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+    If a client was made then an accesses token was successfully reached,
+    therefor the username and password are valid and a connection was made
+    additionally, checks if not using all the optional quota
+    :param client: the client object with an access token
+    :return: ok if got a valid accesses token and not all the quota is used at the moment
+    """
+    _, _, output = check_quota_status_command(client)
+    total = output.get("meta").get("quota").get("total")
+    used = output.get("meta").get("quota").get("used")
+    if total == used:
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:156]: Just to be on the safe side
```suggestion
    if total => used:
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+        return_error(f"Currently using all the optional quota: {used}")
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:157]:
```suggestion
        return_error(f"Quota limitation has been reached: {used}")
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_test.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
+from CrowdStrikeFalconX import Client, test_module,\
+    upload_file_command, send_uploaded_file_to_sendbox_analysis_command, send_url_to_sandbox_analysis_command,\
+    get_full_report_command, get_report_summary_command, get_analysis_status_command, download_ioc_command, \
+    check_quota_status_command, find_sandbox_reports_command, find_submission_id_command
+import pytest
+
+UPLOAD_FILE_ARGS = {
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_test.py:7]: please move the responses and context data to another directory.

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_test.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+            'resources': ['resources1', 'resources2', 'resources3', 'resources4']
+        }
+}
+
+
+@pytest.mark.parametrize('command, args, http_response, context', [
+    (upload_file_command, UPLOAD_FILE_ARGS, UPLOAD_FILE_HTTP_RESPONSE, UPLOAD_FILE_CONTEXT),
+    (send_uploaded_file_to_sendbox_analysis_command, SEND_UPLOADED_FILE_TO_SENDBOX_ANALYSIS_ARGS, SEND_UPLOADED_FILE_TO_SENDBOX_ANALYSIS_HTTP_RESPONSE, SEND_UPLOADED_FILE_TO_SENDBOX_ANALYSIS_CONTEXT),
+    (send_url_to_sandbox_analysis_command, SEND_URL_TO_SANDBOX_ANALYSIS_ARGS, SEND_URL_TO_SANDBOX_ANALYSIS_HTTP_RESPONSE, SEND_URL_TO_SANDBOX_ANALYSIS_CONTEXT),
+    (get_full_report_command, GET_FULL_REPORT_ARGS, GET_FULL_REPORT_HTTP_RESPONSE, get_full_report_context),
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX_test.py:397]: maybe we should have a few examples. a report for a file, report for url, report for a malicious file, etc.
this is to make sure that we map and parse all the data as needed.

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 12]
```diff
@@ -0,0 +1,604 @@
+category: Authentication
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:1]: please change

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 13]
```diff
@@ -0,0 +1,604 @@
...
+- display: Username
+  name: credentials
+  required: true
+  type: 9
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+description: Fully automated malware analysis.
+display: CrowdStrikeFalconX
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:15]:
```suggestion
display: CrowdStrike FalconX
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 14]
```diff
@@ -0,0 +1,604 @@
...
+name: CrowdStrikeFalconX
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:16]:
```suggestion
name: CrowdStrike FalconX
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 15]
```diff
@@ -0,0 +1,604 @@
+category: Authentication
+commonfields:
+  id: CrowdStrikeFalconX
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:3]:
```suggestion
  id: CrowdStrike FalconX
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+
+    response = client.cs_falconx_http_req("POST", url_suffix, content_type='application/octet-stream', data=data)
+
+    resources_fields = ["file_name", "sha256"]
+    entry_context = parse_outputs(response, resources_fields=resources_fields)
+
+    return response, entry_context, response
+
+
+def send_uploaded_file_to_sendbox_analysis_command(
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:188]:
```suggestion
def send_uploaded_file_to_sandbox_analysis_command(
```

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+        super().__init__(base_url=server_url, verify=use_ssl)
+        self._base_url = server_url
+        self._username = username
+        self._password = password
+        self.token = self._generate_token()
+
+    def cs_falconx_http_req(self, method, url_suffix, content_type=None, data=None, json_data=None, param=None):
+        headers = {
+            'Authorization': 'bearer ' + self.token,
+        }
```
Comment (bakatzir) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:47]: move as property

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py (hunk 7 above)
Comment (ohaim1008) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:363]: the merging was tested and works
the context has been changed

**Comment by ohaim1008** [comment]
moved to: https://github.com/demisto/content/pull/7217

**Pull request closed without merging**